builder.go

```diff
@@ -250,27 +255,120 @@ func (k *kResolver) makeAddresses(e EndpointSlice) ([]resolver.Address, string)
 		}
 	}
 
-	return newAddrs, ""
+	return newAddrs
 }
 
-func (k *kResolver) handle(e EndpointSlice) {
-	addrs, _ := k.makeAddresses(e)
-	if len(addrs) > 0 {
+// Aggregate addresses from all endpoint slices
+func (k *kResolver) aggregateAllAddresses() []resolver.Address {
+	k.endpointSlicesMu.RLock()
+	defer k.endpointSlicesMu.RUnlock()
+
+	var allAddresses []resolver.Address
+	totalEndpoints := 0
+
+	for _, slice := range k.endpointSlices {
+		addrs := k.makeAddresses(slice)
+		allAddresses = append(allAddresses, addrs...)
+		totalEndpoints += len(slice.Endpoints)
+	}
+
+	// Update metrics with aggregated data
+	k.endpoints.Set(float64(totalEndpoints))
+	k.addresses.Set(float64(len(allAddresses)))
+
+	return allAddresses
+}
+
+// Handle individual endpoint slice events
+func (k *kResolver) handleEndpointSliceEvent(event Event) {
+	k.endpointSlicesMu.Lock()
+	defer k.endpointSlicesMu.Unlock()
+
+	sliceName := event.Object.Metadata.Name
+	if sliceName == "" {
+		// Fallback: generate name from first endpoint if metadata is missing
+		if len(event.Object.Endpoints) > 0 && len(event.Object.Endpoints[0].Addresses) > 0 {
+			sliceName = fmt.Sprintf("slice-%s", event.Object.Endpoints[0].Addresses[0])
+		} else {
+			sliceName = fmt.Sprintf("slice-%d", len(k.endpointSlices))
+		}
+	}
```

Comment:
Due to the fact that endpoints can be duplicated, I think that using a hashmap with keys based on endpoints might not work.